tests/sentry/release_health/test_tasks.py

```diff
@@ -549,6 +554,85 @@ def test_no_env(self):
+
+        release1 = self.create_release(project=project1, version="foo@1.0.0")
+        release2 = self.create_release(project=project1, version="foo@2.0.0")
+        environment1 = self.create_environment(name="prod", project=project1)
+        environment2 = self.create_environment(name="canary", project=project2)
+
+        self.rpe1 = ReleaseProjectEnvironment.objects.create(
+            project_id=project1.id,
+            release_id=release1.id,
+            environment_id=environment1.id,
+        )
+        self.rpe2 = ReleaseProjectEnvironment.objects.create(
+            project_id=project2.id,
+            release_id=release1.id,
+            environment_id=environment1.id,
+        )
+        self.rpe3 = ReleaseProjectEnvironment.objects.create(
+            project_id=project1.id,
+            release_id=release2.id,
+            environment_id=environment1.id,
+        )
+        self.rpe4 = ReleaseProjectEnvironment.objects.create(
+            project_id=project1.id,
+            release_id=release1.id,
+            environment_id=environment2.id,
+        )
+
+    def test_query_adopted_release_project_environments(self):
+        self.mock_release_project_environments()
+
+        def to_adopted_release(rpe: ReleaseProjectEnvironment) -> AdoptedRelease:
+            return {
+                "project_id": rpe.project_id,
+                "environment": rpe.environment.name,
+                "version": rpe.release.version,
+            }
+
+        # Assert similar releases are ignored. All three pieces of metadata must match
+        # simultaneously for a row to be returned.
+        adopted_releases = [to_adopted_release(self.rpe1)]
+        results = query_adopted_release_project_environments(adopted_releases)
+        assert len(results) == 1
+
+        # Assert only the extra valid release is found.
+        adopted_releases.append(to_adopted_release(self.rpe3))
+        results = query_adopted_release_project_environments(adopted_releases)
+        assert len(results) == 2
+
+        # Extra release not found because it doesn't exist in the database.
+        adopted_releases.append({"project_id": 1, "environment": "any", "version": "whatever"})
+        results = query_adopted_release_project_environments(adopted_releases)
+        assert len(results) == 2
+
+    def test_adopt_release_project_environments(self):
+        self.mock_release_project_environments()
+
+        # Assert none of the releases have been adopted yet.
+        assert self.rpe1.adopted is None
+        assert self.rpe2.adopted is None
+        assert self.rpe3.adopted is None
+        assert self.rpe4.adopted is None
+
+        adopt_release_project_environments([self.rpe1, self.rpe2])
+
+        # Only the releases found in the list are updated. The remainder are unchanged.
+        assert self.rpe1.adopted is not None
+        assert self.rpe2.adopted is not None  # type: ignore[unreachable]
```

Comment:
@asottile-sentry Optional side quest that I think might interest you.  Mypy says this statement is unreachable. At runtime we do make it to the end. Any idea why (or how to debug)?

Reply:
yeah -- in this case you'll either want to save to a variable or rework the assertions

otherwise mypy isn't checking any of this